File: go/internal/testutil/change_files.go
package testutil

import (
	"os"
	"path/filepath"
	"testing"

	"github.com/microsoft/beachball/internal/types"
)

const FakeEmail = "test@test.com"

// WriteChangeFilesFn is the function used by GenerateChangeFiles to write change files.
// Tests must set this to changefile.WriteChangeFiles to break the import cycle.
// Example: testutil.WriteChangeFilesFn = changefile.WriteChangeFiles
var WriteChangeFilesFn func(options *types.BeachballOptions, changes []types.ChangeFileInfo) error

// GetChange creates a ChangeFileInfo with sensible defaults.
// Mirrors the TS getChange fixture helper.
func GetChange(packageName string, comment string, changeType types.ChangeType) types.ChangeFileInfo {
	if comment == "" {
		comment = packageName + " comment"
	}
	if changeType == "" {
		changeType = types.ChangeTypeMinor
	}
	return types.ChangeFileInfo{
		Type:                changeType,
		Comment:             comment,
		PackageName:         packageName,
		Email:               FakeEmail,
		DependentChangeType: types.ChangeTypePatch,
	}
}

// GenerateChanges creates ChangeFileInfo entries from package names with defaults.
// Mirrors the TS generateChanges fixture helper.
func GenerateChanges(packages []string) []types.ChangeFileInfo {
	changes := make([]types.ChangeFileInfo, len(packages))
	for i, pkg := range packages {
		changes[i] = GetChange(pkg, "", "")
	}
	return changes
}

// GenerateChangeFiles creates change files for the given packages using the real
// WriteChangeFiles implementation. Mirrors the TS generateChangeFiles fixture helper.
// WriteChangeFilesFn must be set before calling this function.
func GenerateChangeFiles(t *testing.T, packages []string, options *types.BeachballOptions, repo *Repository) {
	t.Helper()

	if WriteChangeFilesFn == nil {
		t.Fatal("testutil.WriteChangeFilesFn must be set (e.g. testutil.WriteChangeFilesFn = changefile.WriteChangeFiles)")
	}

	changes := make([]types.ChangeFileInfo, len(packages))
	for i, pkg := range packages {
		changes[i] = GetChange(pkg, "test change", types.ChangeTypePatch)
	}

	if err := WriteChangeFilesFn(options, changes); err != nil {
		t.Fatalf("failed to write change files: %v", err)
	}
}

// GetChangeFiles returns the list of change file paths.
func GetChangeFiles(options *types.BeachballOptions) []string {
	changePath := filepath.Join(options.Path, options.ChangeDir)
	entries, err := os.ReadDir(changePath)
	if err != nil {
		return nil
	}

	var files []string
	for _, entry := range entries {
		if filepath.Ext(entry.Name()) == ".json" {
			files = append(files, filepath.Join(changePath, entry.Name()))
		}
	}
	return files
}

File: rust/tests/common/change_files.rs
use beachball::types::change_info::ChangeFileInfo;
use beachball::types::change_info::ChangeType;
use beachball::types::options::BeachballOptions;
use std::path::Path;

use super::repository::Repository;

/// Generate change files for the given packages and commit them.
pub fn generate_change_files(packages: &[&str], options: &BeachballOptions, repo: &Repository) {
    let change_path = Path::new(&options.path).join(&options.change_dir);
    std::fs::create_dir_all(&change_path).ok();

    for pkg in packages {
        let uuid = uuid::Uuid::new_v4();
        let sanitized = pkg.replace(|c: char| !c.is_alphanumeric() && c != '@', "-");
        let filename = format!("{sanitized}-{uuid}.json");
        let file_path = change_path.join(&filename);

        let change = ChangeFileInfo {
            change_type: ChangeType::Patch,
            comment: "test change".to_string(),
            package_name: pkg.to_string(),
            email: "test@test.com".to_string(),
            dependent_change_type: ChangeType::Patch,
        };

        let json = serde_json::to_string_pretty(&change).unwrap();
        std::fs::write(&file_path, json).unwrap();
    }

    repo.git(&["add", "-A"]);
    if options.commit {
        repo.git(&["commit", "-m", "Change files"]);
    }
}

/// Get the list of change file paths for the given options.
pub fn get_change_files(options: &BeachballOptions) -> Vec<String> {
    let change_path = Path::new(&options.path).join(&options.change_dir);
    if !change_path.exists() {
        return vec![];
    }

    let mut files: Vec<String> = Vec::new();
    if let Ok(entries) = std::fs::read_dir(&change_path) {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.extension().and_then(|e| e.to_str()) == Some("json") {
                files.push(path.to_string_lossy().to_string());
            }
        }
    }
    files.sort();
    files
}
